opte-api/src/dns.rs

```diff
@@ -0,0 +1,223 @@
+    const MAX_ENCODED_LEN: usize = 255;
+    const MAX_LABEL_LEN: usize = 63;
+
+    /// Encode the full domain name in the format prescribed by RFC 1035 section
+    /// 3.1.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use opte_api::DomainName;
+    /// let n: DomainName = "quuz.bar.com".parse().unwrap();
+    /// assert_eq!(n.encode(), b"\x04quuz\x03bar\x03com\x00");
+    /// ```
+    pub fn encode(&self) -> &[u8] {
+        self.encoded.as_slice()
+    }
+
+    /// Return the validated domain name string.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+}
+
+impl TryFrom<&[u8]> for DomainName {
+    type Error = &'static str;
+
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        // Sanity check the size of the buffer.
+        //
+        // From RFC 1035, section 3.1, this is limited to 255 octets.
+        if buf.len() > Self::MAX_ENCODED_LEN || buf.is_empty() {
+            return Err("Encoded domain name too long or empty");
+        }
+
+        // Pull out each label length, then the label itself.
+        let mut start: u8 = 0;
+        let mut name = String::new();
+        loop {
+            let len = buf[usize::from(start)];
+
+            // Root label has zero length.
+            if len == 0 {
+                break;
+            }
+
+            // The name starts on the next octet. Check that the end is within a
+            // u8 and the length of the string!
+            start += 1;
+            let Some(end) = start.checked_add(len) else {
+                return Err("Invalid label length");
+            };
+            let Some(chunk) = buf.get(usize::from(start)..usize::from(end)) else {
+                return Err("Invalid label length");
+            };
+
+            // Push the label and _then_ the label-separator.
+            let Ok(label) = str::from_utf8(chunk) else {
+                return Err("Non-UTF8 label");
+            };
+            name.push_str(label);
+            name.push('.');
+
+            // Off to the next label.
+            start += len;
+        }
+
+        // We've already pushed the root label separator, since we push `'.'`
+        // after each label, including the last, before breaking.
+        Ok(Self { encoded: buf.to_vec(), name })
+    }
+}
+
+impl FromStr for DomainName {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.len() > Self::MAX_STR_LEN || s.is_empty() {
+            return Err("Domain name too long or empty");
+        }
+        if s.starts_with('.') {
+            return Err("Domain name cannot start with '.'");
+        }
+
+        // Pull out a possible `.` at the very end, and handle that separately.
+        let is_fqdn = s.ends_with('.');
+        let relative_name = if is_fqdn { &s[..s.len() - 1] } else { s };
+
+        // Split out each label, and create the encoded form.
+        let mut encoded = Vec::with_capacity(s.len());
```

Comment:
Maybe? I don't think it's always +2, though. If the name ends in a `.`, then the length already includes the root label length of zero. I thought it'd be more complicated, but it could just be `s.len() + 1` plus 1 if `s` isn't fully-qualified. I'll do that now, thanks!

Reply:
Oh, that might be why I had `is_fqdn` :)